### DISCUSSION TITLE
Bind to Unix Socket

### DISCUSSION THREAD

**Discussion by mikehu404** [meilisearch/product#622] [Feedback & Feature Proposal]
Hello,
I wonder if meilisearch server could bind to unix socket,
I can't find it anywhere in meilisearch's doc.
Unix socket is faster than tcp, as discussed [here](https://stackoverflow.com/a/29436429)
It would be great to have this feature.
Especially if you want to put meilisearch behind a reverse proxy like nginx or haproxy.

**Comment by gmourier** [comment]
Hello @mikehjc

Thanks for sharing that!

We won't work on it because we believe that the impact on performance will not be significant since:

- Meilisearch opens a lot of internal dbs, and the time you save on packet transfer will always be lessened by it. The order of magnitude is negligible.
- Regarding document updates, the indexing process affects multiple structures on disks and will significantly reduce the impact of this optimization.

Let me know if it's unclear

cc @irevoire @Kerollmops

**Reply by michail-vestnik** [reply]
Hey. I'm not clear.
I don't understand what internal databases and disk operations have to do with the transfer of data from CLIENT to MEILISEARCH.
For example, in Redis unix sockets allow to speed up data transfer by about 20%.

**Reply by gmourier** [reply]
Hey @michail-vestnik

Even if the UNIX sockets help to gain a % of the speed in data transferring, you will see few enhancements regarding the indexing speed since the bottleneck in a great majority of cases is situated at the computation of the different structures when the data arrives at the destination.

If it's unclear, let me know, so I can ping someone more technical than me to give more explanation. Thanks!

**Reply by mikehu404** [reply]
> Hello @gmourier

Yes, The performance gain will be not be significant.

But i still believe it's beneficial to have the ability to bind to unix socket:
- If I were to put meilisearch behind a reverse proxy like nginx or haproxy, and have the reverse proxy deal with TLS offloading and load balancing. In terms of performance, using Unix socket is better that using TCP to connect reverse proxy and meilisearch.
- unix socket have the ability to set permissions, which is good for security.

**Reply by Kerollmops** [reply]
Hey @mikehu404 👋

[Someone opened a PR to support this](https://github.com/meilisearch/meilisearch/pull/4528), and everything seems to be handled by Actix Web, our HTTP Server dependency. If you want to take a look, propose a better approach, or even finish the PR, it would be great 😉

Have a nice day 🎠
